server/src/main/java/org/opensearch/cluster/routing/allocation/decider/ThrottlingAllocationDecider.java

```diff
@@ -150,36 +165,73 @@ public Decision canAllocate(ShardRouting shardRouting, RoutingNode node, Routing
+                replicasInitialRecoveries),
+            String.format(Locale.ROOT, "[%s=%d]", CLUSTER_ROUTING_ALLOCATION_NODE_INITIAL_REPLICAS_RECOVERIES_SETTING.getKey(),
+                replicasInitialRecoveries));
+    }
+
+    private Decision allocateNonIntialReplicas(ShardRouting shardRouting, RoutingNode node, RoutingAllocation allocation) {
+        int currentInRecoveries = allocation.routingNodes().getIncomingRecoveries(node.nodeId());
+
+        return allocateReplicas(shardRouting, allocation, currentInRecoveries, concurrentIncomingRecoveries,
+            (x,y) -> getPrimaryNodeOutgoingRecoveries(x,y), concurrentOutgoingRecoveries,
+            String.format(Locale.ROOT, "[%s=%d] (can also be set via [%s])",
+                CLUSTER_ROUTING_ALLOCATION_NODE_CONCURRENT_INCOMING_RECOVERIES_SETTING.getKey(),
+            concurrentIncomingRecoveries, CLUSTER_ROUTING_ALLOCATION_NODE_CONCURRENT_RECOVERIES_SETTING.getKey()),
+            String.format(Locale.ROOT, "[%s=%d] (can also be set via [%s])",
+                CLUSTER_ROUTING_ALLOCATION_NODE_CONCURRENT_OUTGOING_RECOVERIES_SETTING.getKey(),
+                concurrentOutgoingRecoveries, CLUSTER_ROUTING_ALLOCATION_NODE_CONCURRENT_RECOVERIES_SETTING.getKey()));
+    }
+
+    private Integer getPrimaryNodeOutgoingRecoveries(ShardRouting shardRouting, RoutingAllocation allocation) {
+        ShardRouting primaryShard = allocation.routingNodes().activePrimary(shardRouting.shardId());
+        return allocation.routingNodes().getOutgoingRecoveries(primaryShard.currentNodeId());
+    }
+
+    private Integer getInitialPrimaryNodeOutgoingRecoveries(ShardRouting shardRouting, RoutingAllocation allocation) {
+        ShardRouting primaryShard = allocation.routingNodes().activePrimary(shardRouting.shardId());
+        return allocation.routingNodes().getInitialOutgoingRecoveries(primaryShard.currentNodeId());
+    }
+
+    private Decision allocateReplicas(ShardRouting shardRouting, RoutingAllocation allocation, int currentInRecoveries,
+                                      int inRecoveriesLimit, BiFunction<ShardRouting, RoutingAllocation,
+                                      Integer> primaryNodeOutRecoveriesFunc, int outRecoveriesLimit,
+                                      String incomingRecoveriesSettingMsg, String outGoingRecoveriesSettingMsg) {
+        // Allocating a shard to this node will increase the incoming recoveries
+        if (currentInRecoveries >= inRecoveriesLimit) {
+            return allocation.decision(THROTTLE, NAME,
+                "reached the limit of incoming shard recoveries [%d], cluster setting %s",
+                currentInRecoveries, incomingRecoveriesSettingMsg);
+        } else {
+            // search for corresponding recovery source (= primary shard) and check number of outgoing recoveries on that node
+            ShardRouting primaryShard = allocation.routingNodes().activePrimary(shardRouting.shardId());
+            if (primaryShard == null) {
```

Comment:
Add `primaryShard.isActive() == false` as well?

Reply:
The field will be null if primary is not allocated.